llvm/lib/Target/X86/X86KulaevIncremCounterPass.cpp

```diff
@@ -0,0 +1,72 @@
+#include "X86.h"
+#include "X86InstrInfo.h"
+#include "X86Subtarget.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+
+using namespace llvm;
+
+#define PASS_NAME "X86KulaevIncremCounterPass"
+
+namespace {
+class X86KulaevIncremCounterPass : public MachineFunctionPass {
+public:
+  static char ID;
+  X86KulaevIncremCounterPass() : MachineFunctionPass(ID) {}
+
+  bool runOnMachineFunction(MachineFunction &MachineFunctionRunning) override;
+
+private:
+  void addInReg(MachineFunction &mi, DebugLoc dl, const TargetInstrInfo *TII,
+                unsigned registreIc);
+  void updateCount(DebugLoc dl, MachineFunction &mf, const TargetInstrInfo *ti,
+                   unsigned registreIc);
+};
+} // namespace
+
+bool X86KulaevIncremCounterPass::runOnMachineFunction(
+    MachineFunction &MachineFunctionRunning) {
+  const TargetInstrInfo *TargetInstructionInfo =
+      MachineFunctionRunning.getSubtarget().getInstrInfo();
+  DebugLoc DL3 = MachineFunctionRunning.front().begin()->getDebugLoc();
+
+  unsigned icReg = MachineFunctionRunning.getRegInfo().createVirtualRegister(
+      &X86::GR64RegClass);
+
+  updateCount(DL3, MachineFunctionRunning, TargetInstructionInfo, icReg);
+  addInReg(MachineFunctionRunning, DL3, TargetInstructionInfo, icReg);
+
+  return true;
+}
+
+void X86KulaevIncremCounterPass::updateCount(DebugLoc dl, MachineFunction &mf,
+                                             const TargetInstrInfo *ti,
+                                             unsigned registreIc) {
+  unsigned count = 0;
+  for (auto &MBasicBlock : mf) {
+    for (auto &MInstruction : MBasicBlock) {
+      ++count;
```

Comment:
This is just a number of instructions in assembler code. Whilst number of executed instructions during a program run is required. During execution some of blocks may be executed many times if they are in a loop.

Reply:
@m-ly4 
As I understand it, do I need to multiply the number of instructions in the basic block, which is a cycle, by the number of passes in this cycle? Can you please tell me how I could implement this? Thank you in advance!

Reply:
You don't have to account number of loop iteration. It may be known dynamically depending on different input parameters.
Simpler solution is just update counter in each block and don't forget to save it into memory before any return from function.